src/sdk/pynni/nni/compression/torch/utils/counter.py

```diff
@@ -0,0 +1,103 @@
+# Copyright (c) Microsoft Corporation.
+# Licensed under the MIT license.
+
+import torch
+import torch.nn as nn
+from torch.nn.modules.conv import _ConvNd
+from torch.nn.modules.batchnorm import _BatchNorm
+from nni.compression.torch.compressor import PrunerModuleWrapper
+import logging
+
+_logger = logging.getLogger(__name__)
+
+try:
+    from thop import profile
+except ImportError:
+    _logger.warning('Please install thop using command: pip install thop')
+    
+def count_flops_params(model: nn.Module, input_size=None, verbose=True):
+    """
+    Count FLOPs and Params of the given model. 
+    This function would identify the mask on the module 
+    and take the pruned shape into consideration. 
+    Note that, for sturctured pruning, we only identify 
+    the remained filters according to its mask, which 
+    not taking the pruned input channels into consideration,
+    so the calculated FLOPs will be larger than real number.
+
+    Parameters
+    ---------
+    model : nn.Module
+        target model.
+    input_size: list, tuple
+        the input shape of data
+
+
+    Returns
+    -------
+    flops: float
+        total flops of the model
+    params:
+        total params of the model
+    """
+
+    assert input_size is not None
+
+    device = next(model.parameters()).device
+    inputs = torch.randn(input_size).to(device)
+
+    hook_module_list = []
+    prev_m = None
+    for m in model.modules():
+        weight_mask = None
+        m_type = type(m)
+        if m_type in custom_ops:
+            if isinstance(prev_m, PrunerModuleWrapper):
+                weight_mask = prev_m.weight_mask
+
+            m.register_buffer('weight_mask', weight_mask)
+            hook_module_list.append(m)
+        prev_m = m
+
+    flops, params = profile(model, inputs=(inputs, ), custom_ops=custom_ops, verbose=verbose)
+
+    for m in hook_module_list:
+        m._buffers.pop("weight_mask")
+
+    return flops, params
+
+def count_convNd_mask(m, x, y):
+    output_channel = y.size()[1]
+    output_size =  torch.zeros(y.size()[2:]).numel()
+    kernel_size = torch.zeros(m.weight.size()[2:]).numel()
+
+    bias_flops = 1 if m.bias is not None else 0
+
+    if m.weight_mask is not None:
+        output_channel = m.weight_mask.sum() // (m.in_channels * kernel_size)
+
+    total_ops = output_channel * output_size * (m.in_channels // m.groups * kernel_size + bias_flops)
```

Comment:
does `total_ops` mean total flops?

Reply:
Yes. Because [thop#L73](https://github.com/Lyken17/pytorch-OpCounter/blob/dbc052ec2d914e950b3e0fa24d3eea753a4e0bb7/thop/profile.py#L73) register the initial state with `total_ops`.

Reply:
got it